File: crates/dashboard-model/src/victron_controller/dashboard/ws_client_message.rs
use crate::victron_controller::dashboard::command::Command;
use crate::victron_controller::dashboard::ws_ping::WsPing;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct Ping {
    pub body: WsPing,
}



#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct SendCommand {
    pub body: Command,
}



impl crate::baboon_runtime::BaboonBinCodecIndexed for Ping {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        1
    }
}

impl crate::baboon_runtime::BaboonBinEncode for Ping {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                value.body.encode_ueba(ctx, &mut buffer)?;
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            value.body.encode_ueba(ctx, writer)?;
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for Ping {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let body = WsPing::decode_ueba(ctx, reader)?;
        Ok(Ping {
            body,
        })
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for SendCommand {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        1
    }
}

impl crate::baboon_runtime::BaboonBinEncode for SendCommand {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                value.body.encode_ueba(ctx, &mut buffer)?;
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            value.body.encode_ueba(ctx, writer)?;
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for SendCommand {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let body = Command::decode_ueba(ctx, reader)?;
        Ok(SendCommand {
            body,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum WsClientMessage {
    Ping(Ping),
    SendCommand(SendCommand),
}

impl serde::Serialize for WsClientMessage {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        use serde::ser::SerializeMap;
        let mut map = serializer.serialize_map(Some(1))?;
        match self {
            WsClientMessage::Ping(v) => {
                map.serialize_entry("Ping", v)?;
            }
            WsClientMessage::SendCommand(v) => {
                map.serialize_entry("SendCommand", v)?;
            }
        }
        map.end()
    }
}

impl<'de> serde::Deserialize<'de> for WsClientMessage {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct AdtVisitor;
        impl<'de> serde::de::Visitor<'de> for AdtVisitor {
            type Value = WsClientMessage;
            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(f, "a single-key map representing WsClientMessage")
            }
            fn visit_map<A: serde::de::MapAccess<'de>>(self, mut map: A) -> Result<Self::Value, A::Error> {
                let key: String = map.next_key()?
                    .ok_or_else(|| serde::de::Error::custom("expected single-key map for ADT"))?;
                match key.as_str() {
                    "Ping" => Ok(WsClientMessage::Ping(map.next_value()?)),
                    "SendCommand" => Ok(WsClientMessage::SendCommand(map.next_value()?)),
                    _ => Err(serde::de::Error::unknown_variant(&key, &["Ping", "SendCommand"])),
                }
            }
        }
        deserializer.deserialize_map(AdtVisitor)
    }
}

impl std::fmt::Display for WsClientMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            WsClientMessage::Ping(v) => write!(f, "WsClientMessage::Ping({:?})", v),
            WsClientMessage::SendCommand(v) => write!(f, "WsClientMessage::SendCommand({:?})", v),
        }
    }
}

impl std::error::Error for WsClientMessage {}

impl crate::baboon_runtime::BaboonBinCodecIndexed for WsClientMessage {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for WsClientMessage {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            WsClientMessage::Ping(v) => {
                crate::baboon_runtime::bin_tools::write_byte(writer, 0)?;
                v.encode_ueba(ctx, writer)?;
            }
            WsClientMessage::SendCommand(v) => {
                crate::baboon_runtime::bin_tools::write_byte(writer, 1)?;
                v.encode_ueba(ctx, writer)?;
            }
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for WsClientMessage {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => {
                let v = Ping::decode_ueba(ctx, reader)?;
                Ok(WsClientMessage::Ping(v))
            }
            1 => {
                let v = SendCommand::decode_ueba(ctx, reader)?;
                Ok(WsClientMessage::SendCommand(v))
            }
            _ => Err(format!("Unknown ADT branch tag: {}", tag).into()),
        }
    }
}

File: web/src/model/victron_controller/dashboard/v0_2_0/WsClientMessage.ts
// @ts-nocheck
import {BaboonGenerated, BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../../BaboonSharedRuntime'
import {WsPing, WsPing_UEBACodec} from './WsPing'
import {Command, Command_UEBACodec} from './Command'

export type WsClientMessage = Ping | SendCommand

export const WsClientMessage = {
    BaboonDomainVersion: '0.2.0',
    BaboonDomainIdentifier: 'victron_controller.dashboard',
    BaboonTypeIdentifier: 'victron_controller.dashboard/:#WsClientMessage',
    BaboonSameInVersions: ["0.2.0", "0.3.0"],
    BaboonAdtTypeIdentifier: 'victron_controller.dashboard/:#WsClientMessage',
    binCodec(): WsClientMessage_UEBACodec {
        return WsClientMessage_UEBACodec.instance
    }
} as const

export function isPing(value: WsClientMessage): value is Ping { return value instanceof Ping; }
export function isSendCommand(value: WsClientMessage): value is SendCommand { return value instanceof SendCommand; }

export class Ping implements BaboonGenerated {
    private readonly _body: WsPing;

    constructor(body: WsPing) {
        this._body = body
    }

    public get body(): WsPing {
        return this._body;
    }

    public toJSON(): Record<string, unknown> {
        return {
            body: this._body
        };
    }

    public with(overrides: {body?: WsPing}): Ping {
        return new Ping(
            'body' in overrides ? overrides.body! : this._body
        );
    }

    public static fromPlain(obj: {body: WsPing}): Ping {
        return new Ping(
            obj.body
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return Ping.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return Ping.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#Ping'
    public baboonTypeIdentifier() {
        return Ping.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.1.0", "0.2.0", "0.3.0"]
    public baboonSameInVersions() {
        return Ping.BaboonSameInVersions
    }
    public static readonly BaboonAdtTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#Ping'
    public baboonAdtTypeIdentifier() {
        return Ping.BaboonAdtTypeIdentifier
    }
    
    public static binCodec(): Ping_UEBACodec {
        return Ping_UEBACodec.instance
    }
}

/** @deprecated Version 0.2.0 is deprecated, you should migrate to 0.3.0 */
export class Ping_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: Ping, writer: BaboonBinWriter): unknown {
        if (this !== Ping_UEBACodec.lazyInstance.value) {
          return Ping_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                WsPing_UEBACodec.instance.encode(ctx, value.body, buffer);
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            WsPing_UEBACodec.instance.encode(ctx, value.body, writer);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): Ping {
        if (this !== Ping_UEBACodec .lazyInstance.value) {
            return Ping_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 1; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const body = WsPing_UEBACodec.instance.decode(ctx, reader);
        return new Ping(
            body,
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return Ping_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return Ping_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#Ping'
    public baboonTypeIdentifier() {
        return Ping_UEBACodec.BaboonTypeIdentifier
    }
    public static readonly BaboonAdtTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#Ping'
    public baboonAdtTypeIdentifier() {
        return Ping_UEBACodec.BaboonAdtTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new Ping_UEBACodec())
    public static get instance(): Ping_UEBACodec {
        return Ping_UEBACodec.lazyInstance.value
    }
}

export class SendCommand implements BaboonGenerated {
    private readonly _body: Command;

    constructor(body: Command) {
        this._body = body
    }

    public get body(): Command {
        return this._body;
    }

    public toJSON(): Record<string, unknown> {
        return {
            body: this._body
        };
    }

    public with(overrides: {body?: Command}): SendCommand {
        return new SendCommand(
            'body' in overrides ? overrides.body! : this._body
        );
    }

    public static fromPlain(obj: {body: Command}): SendCommand {
        return new SendCommand(
            obj.body
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return SendCommand.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return SendCommand.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#SendCommand'
    public baboonTypeIdentifier() {
        return SendCommand.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.2.0", "0.3.0"]
    public baboonSameInVersions() {
        return SendCommand.BaboonSameInVersions
    }
    public static readonly BaboonAdtTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#SendCommand'
    public baboonAdtTypeIdentifier() {
        return SendCommand.BaboonAdtTypeIdentifier
    }
    
    public static binCodec(): SendCommand_UEBACodec {
        return SendCommand_UEBACodec.instance
    }
}

/** @deprecated Version 0.2.0 is deprecated, you should migrate to 0.3.0 */
export class SendCommand_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: SendCommand, writer: BaboonBinWriter): unknown {
        if (this !== SendCommand_UEBACodec.lazyInstance.value) {
          return SendCommand_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                Command_UEBACodec.instance.encode(ctx, value.body, buffer);
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            Command_UEBACodec.instance.encode(ctx, value.body, writer);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): SendCommand {
        if (this !== SendCommand_UEBACodec .lazyInstance.value) {
            return SendCommand_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 1; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const body = Command_UEBACodec.instance.decode(ctx, reader);
        return new SendCommand(
            body,
        );
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return SendCommand_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return SendCommand_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#SendCommand'
    public baboonTypeIdentifier() {
        return SendCommand_UEBACodec.BaboonTypeIdentifier
    }
    public static readonly BaboonAdtTypeIdentifier = 'victron_controller.dashboard/[victron_controller.dashboard/:#WsClientMessage]#SendCommand'
    public baboonAdtTypeIdentifier() {
        return SendCommand_UEBACodec.BaboonAdtTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new SendCommand_UEBACodec())
    public static get instance(): SendCommand_UEBACodec {
        return SendCommand_UEBACodec.lazyInstance.value
    }
}


/** @deprecated Version 0.2.0 is deprecated, you should migrate to 0.3.0 */
export class WsClientMessage_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: WsClientMessage, writer: BaboonBinWriter): unknown {
        if (this !== WsClientMessage_UEBACodec.lazyInstance.value) {
          return WsClientMessage_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (value instanceof Ping) {
                BinTools.writeByte(writer, 0);
                Ping_UEBACodec.instance.encode(ctx, value, writer);
            }
            if (value instanceof SendCommand) {
                BinTools.writeByte(writer, 1);
                SendCommand_UEBACodec.instance.encode(ctx, value, writer);
            }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): WsClientMessage {
        if (this !== WsClientMessage_UEBACodec .lazyInstance.value) {
            return WsClientMessage_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return Ping_UEBACodec.instance.decode(ctx, reader)
                case 1: return SendCommand_UEBACodec.instance.decode(ctx, reader)
            default: throw new Error("Unknown ADT branch tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.2.0'
    public baboonDomainVersion() {
        return WsClientMessage_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return WsClientMessage_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#WsClientMessage'
    public baboonTypeIdentifier() {
        return WsClientMessage_UEBACodec.BaboonTypeIdentifier
    }
    public static readonly BaboonAdtTypeIdentifier = 'victron_controller.dashboard/:#WsClientMessage'
    public baboonAdtTypeIdentifier() {
        return WsClientMessage_UEBACodec.BaboonAdtTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new WsClientMessage_UEBACodec())
    public static get instance(): WsClientMessage_UEBACodec {
        return WsClientMessage_UEBACodec.lazyInstance.value
    }
}